sourced/ml/core/models/id_splitter.py

```diff
@@ -0,0 +1,121 @@
+import logging
+import string
+from typing import List, Sequence, Tuple
+
+import keras
+from keras.preprocessing.sequence import pad_sequences
+from modelforge import Model, register_model
+import numpy
+
+from sourced.ml.core.algorithms.id_splitter.nn_model import (f1score, precision,
+                                                             recall)
+from sourced.ml.core.models.license import DEFAULT_LICENSE
+
+
+@register_model
+class IdentifierSplitterNN(Model):
+    """
+    Bi-Directionnal LSTM Model. Splits identifiers without need for a conventional pattern.
+    """
+    NAME = "id_splitter_nn"
+    VENDOR = "source{d}"
+    DESCRIPTION = "Weights of the BiLSTM network to split source code identifiers."
+    LICENSE = DEFAULT_LICENSE
+
+    DEFAULT_MAXLEN = 40
+    DEFAULT_PADDING = "post"
+    DEFAULT_MAPPING = None
+
+    def construct(self, model: "keras.models.Model",
+                  maxlen: int = DEFAULT_MAXLEN,
+                  padding: str = DEFAULT_PADDING,
+                  mapping: dict = DEFAULT_MAPPING) -> "IdentifierSplitterNN":
+
+        self._maxlen = maxlen
+        self._padding = padding
+        self._mapping = mapping
+        self._model = model
+        return self
+
+    @property
+    def model(self) -> "keras.models.Model":
+        """
+        Returns the wrapped keras model.
+        """
+        return self._model
+
+    def clear_keras(self):
+        """
+        Clears the keras session used to run the model
+        """
+        keras.backend.clear_session()
+
+    def _generate_tree(self) -> dict:
+        return {
+            "config": self._model.get_config(),
+            "weights": self._model.get_weights(),
+            "mapping": self._mapping,
+            }
+
+    def _load_tree(self, tree: dict):
+        model = keras.models.Model.from_config(tree["config"])
+        model.set_weights(tree["weights"])
+        if "mapping" in tree:
+            self.construct(model, mapping=tree["mapping"])
+        else:
+            self.construct(model)
+
+    def _prepare_single_identifier(self, identifier: str) -> Tuple[numpy.array, str]:
+        if self._mapping is None:
+            self._mapping = {c: i for i, c in enumerate(string.ascii_lowercase, start=1)}
+
+        # Clean identifier
+        clean_id = "".join(char for char in identifier.lower() if char in self._mapping)
+        if len(clean_id) > self._maxlen:
+            clean_id = clean_id[:self._maxlen]
+        logging.info("Preprocessed identifier: {}".format(clean_id))
```

Comment:
Define a class-level `_log = logging.getLogger("IdentifierSplitterBiLSTM")` and use it for logging.

Reply:
The log level should be debug

Reply:
We use only `%` ("old") style string substitutions. https://github.com/src-d/guide/blob/master/engineering/conventions/python.md#code-style

Reply:
Please include the original string into the log message